rust/cadeau/src/xmf/vpx/mod.rs

```diff
@@ -0,0 +1,225 @@
+use core::fmt;
+
+use xmf_sys::{
+    vpx::{XmfVpxCodecType, XmfVpxPacketKind},
+    XmfVpxFrame, XmfVpxFrame_Destroy, XmfVpxFrame_GetBuffer, XmfVpxFrame_GetDuration, XmfVpxFrame_GetFlags,
+    XmfVpxFrame_GetHeight, XmfVpxFrame_GetPartitionId, XmfVpxFrame_GetPts, XmfVpxFrame_GetSize,
+    XmfVpxFrame_GetSpatialLayerEncoded, XmfVpxFrame_GetWidth, XmfVpxImage, XmfVpxImage_Destroy, XmfVpxPacket,
+    XmfVpxPacket_Destroy, XmfVpxPacket_GetFrame, XmfVpxPacket_GetKind, XmfVpxPacket_IsEmpty,
+};
+
+pub mod decoder;
+pub mod encoder;
+
+#[derive(Debug, Clone, Copy)]
+pub enum VpxCodec {
+    VP8,
+    VP9,
+}
+
+pub struct VpxImage {
+    pub(crate) ptr: *const XmfVpxImage,
+}
+
+impl Drop for VpxImage {
+    fn drop(&mut self) {
+        // Safety: it is safe to call, the owenership of the pointer is managed by the XmfVpxImage itself.
+        unsafe {
+            XmfVpxImage_Destroy(self.ptr.cast_mut());
+        }
+    }
+}
+
+impl From<VpxCodec> for XmfVpxCodecType {
+    fn from(codec: VpxCodec) -> Self {
+        match codec {
+            VpxCodec::VP8 => XmfVpxCodecType::VP8,
+            VpxCodec::VP9 => XmfVpxCodecType::VP9,
+        }
+    }
+}
+
+pub fn is_key_frame(buffer: &[u8]) -> bool {
+    if buffer.is_empty() {
+        return false;
+    }
+    buffer[0] & 0x1 == 0
+}
+
+pub struct VpxPacket {
+    pub(crate) ptr: *const XmfVpxPacket,
+}
+
+impl VpxPacket {
+    /// # Safety
+    ///
+    /// This function is `unsafe` because it assumes that `self.ptr` is a valid pointer.
+    /// It should not be used after any call to a `VpxEncoder` method that modifies or releases the packet.
+    pub unsafe fn kind(&self) -> XmfVpxPacketKind {
+        // Safety: It's up to the caller to ensure the packet is valid.
+        unsafe { XmfVpxPacket_GetKind(self.ptr.cast_mut()) }
```

Comment:
Same issue as above. Re-check for this pattern.

Reply:
updated